runners/direct-java/src/main/java/org/apache/beam/runners/direct/ParDoEvaluatorFactory.java

```diff
@@ -99,13 +117,26 @@ public void cleanup() throws Exception {
       CommittedBundle<InputT> inputBundle)
       throws Exception {
     String stepName = evaluationContext.getStepName(application);
-    DirectStepContext stepContext =
+    final DirectStepContext stepContext =
         evaluationContext
             .getExecutionContext(application, inputBundle.getKey())
             .getOrCreateStepContext(stepName, stepName);
 
-    DoFnLifecycleManager fnManager =
-        fnClones.getUnchecked(hooks.getDoFn(application.getTransform()));
+    final DoFn<?, ?> doFn = hooks.getDoFn(application.getTransform());
+    DoFnLifecycleManager fnManager = fnClones.getUnchecked(doFn);
+    final DoFnSignature signature = DoFnSignatures.getSignature(doFn.getClass());
+
+    // If the DoFn is stateful, schedule state clearing idempotently
```

Comment:
Done. Added `weakValues` since that also expresses the fact that it is obsolete once the callback is not held elsewhere. This is redundant with the fact that it invalidates the key, but harmless and expressive.